// @ui5/webcomponents-core/dist/sap/ui/core/dnd/DragDropInfo.js
import DragInfo from './DragInfo.js';
import DropInfo from './DropInfo.js';
import Log from '../../../base/Log.js';
var DragDropInfo = DropInfo.extend('sap.ui.core.dnd.DragDropInfo', {
    metadata: {
        library: 'sap.ui.core',
        interfaces: [
            'sap.ui.core.dnd.IDragInfo',
            'sap.ui.core.dnd.IDropInfo'
        ],
        properties: {
            sourceAggregation: {
                type: 'string',
                defaultValue: null
            }
        },
        associations: {
            targetElement: {
                type: 'sap.ui.core.Element',
                multiple: false
            }
        },
        events: {
            dragStart: { allowPreventDefault: true },
            dragEnd: {}
        }
    }
});
DragDropInfo.prototype.isDraggable = DragInfo.prototype.isDraggable;
DragDropInfo.prototype.fireDragEnd = DragInfo.prototype.fireDragEnd;
DragDropInfo.prototype.fireDragStart = DragInfo.prototype.fireDragStart;
DragDropInfo.prototype.getDropTarget = function () {
    var sTargetElement = this.getTargetElement();
    if (sTargetElement) {
        return sap.ui.getWCCore().byId(sTargetElement);
    }
    return this.getParent();
};
DragDropInfo.prototype.setGroupName = function () {
    Log.error('groupName property must not be set on ' + this);
    return this;
};
export default DragDropInfo;
