// @ui5/webcomponents-core/dist/sap/ui/core/routing/Views.js
import TargetCache from './TargetCache.js';
import UriParameters from '../../../base/util/UriParameters.js';
import Log from '../../../base/Log.js';
var Views = TargetCache.extend('sap.ui.core.routing.Views', {
    metadata: {
        publicMethods: [
            'getView',
            'setView'
        ]
    },
    constructor: function (oOptions) {
        if (!oOptions) {
            oOptions = {};
        }
        function checkUrl() {
            if (new UriParameters(window.location.href).get('sap-ui-xx-asyncRouting') === 'true') {
                Log.warning('Activation of async view loading in routing via url parameter is only temporarily supported and may be removed soon', 'TargetCache');
                return true;
            }
            return false;
        }
        if (oOptions.async === undefined) {
            oOptions.async = checkUrl();
        }
        TargetCache.apply(this, [oOptions]);
    },
    getView: function (oOptions) {
        return this.get(oOptions, 'View');
    },
    setView: function (sName, oView) {
        return this.set(sName, 'View', oView);
    },
    fireCreated: function (mArguments) {
        if (mArguments) {
            mArguments.view = mArguments.object;
            mArguments.viewOptions = mArguments.options;
        }
        return this.fireEvent('created', mArguments);
    }
});
export default Views;
