// @ui5/webcomponents-core/dist/sap/ui/core/CustomData.js
import Element from './Element.js';
import library from './library.js';
import F6Navigation from '../events/F6Navigation.js';
import Log from '../../base/Log.js';
var ID = library.ID;
var CustomData = Element.extend('sap.ui.core.CustomData', {
    metadata: {
        library: 'sap.ui.core',
        properties: {
            key: {
                type: 'string',
                group: 'Data',
                defaultValue: null
            },
            value: {
                type: 'any',
                group: 'Data',
                defaultValue: null
            },
            writeToDom: {
                type: 'boolean',
                group: 'Data',
                defaultValue: false
            }
        },
        designtime: 'sap/ui/core/designtime/CustomData.designtime'
    }
});
CustomData.prototype.setValue = function (oValue) {
    this.setProperty('value', oValue, true);
    var oControl = this.getParent();
    if (oControl && oControl.getDomRef()) {
        var oCheckResult = this._checkWriteToDom(oControl);
        if (oCheckResult) {
            oControl.$().attr(oCheckResult.key, oCheckResult.value);
        }
    }
    return this;
};
CustomData.prototype._checkWriteToDom = function (oRelated) {
    if (!this.getWriteToDom()) {
        return null;
    }
    var key = this.getKey();
    var value = this.getValue();
    if (typeof value != 'string') {
        Log.error('CustomData with key ' + key + ' should be written to HTML of ' + oRelated + ' but the value is not a string.');
        return null;
    }
    if (!ID.isValid(key) || key.indexOf(':') != -1) {
        Log.error('CustomData with key ' + key + ' should be written to HTML of ' + oRelated + ' but the key is not valid (must be a valid sap.ui.core.ID without any colon).');
        return null;
    }
    if (key == F6Navigation.fastNavigationKey) {
        value = /^\s*(x|true)\s*$/i.test(value) ? 'true' : 'false';
    } else if (key.indexOf('sap-ui') == 0) {
        Log.error('CustomData with key ' + key + ' should be written to HTML of ' + oRelated + ' but the key is not valid (may not start with \'sap-ui\').');
        return null;
    }
    return {
        key: 'data-' + key,
        value: value
    };
};
export default CustomData;
