// @ui5/webcomponents-core/dist/sap/ui/core/XMLComposite.js
import jQuery from '../thirdparty/jquery.js';
import Control from './Control.js';
import XMLCompositeMetadata from './XMLCompositeMetadata.js';
import ManagedObjectModel from '../model/base/ManagedObjectModel.js';
import JSONModel from '../model/json/JSONModel.js';
import Fragment from './Fragment.js';
import ManagedObject from '../base/ManagedObject.js';
import DataType from '../base/DataType.js';
import ResourceModel from '../model/resource/ResourceModel.js';
import Log from '../../base/Log.js';
import Measurement from '../performance/Measurement.js';
var sXMLComposite = 'sap.ui.core.XMLComposite';
var XMLComposite = Control.extend('sap.ui.core.XMLComposite', {
    metadata: {
        interfaces: ['sap.ui.core.IDScope'],
        properties: {
            width: {
                type: 'sap.ui.core.CSSSize',
                group: 'Dimension',
                defaultValue: '100%',
                invalidate: true
            },
            height: {
                type: 'sap.ui.core.CSSSize',
                group: 'Dimension',
                defaultValue: null,
                invalidate: true
            },
            displayBlock: {
                type: 'boolean',
                group: 'Appearance',
                defaultValue: true,
                invalidate: true
            }
        },
        aggregations: {
            _content: {
                type: 'sap.ui.core.Control',
                multiple: false,
                visibility: 'hidden',
                invalidate: true
            }
        }
    },
    constructor: function () {
        this._bIsCreating = true;
        Control.apply(this, arguments);
        delete this._bIsCreating;
    },
    renderer: function (oRm, oControl) {
        Log.debug('Start rendering \'' + oControl.sId, sXMLComposite);
        Measurement.start(oControl.getId() + '---renderControl', 'Rendering of ' + oControl.getMetadata().getName(), [
            'rendering',
            'control'
        ]);
        oRm.write('<div');
        oRm.writeControlData(oControl);
        oRm.writeAccessibilityState(oControl);
        if (!oControl.getDisplayBlock() && (oControl.getWidth() !== '100%' || oControl.getHeight() !== '100%')) {
            oRm.addStyle('display', 'inline-block');
        }
        oRm.writeClasses();
        if (oControl.getHeight()) {
            oRm.addStyle('height', oControl.getHeight());
        }
        if (oControl.getWidth()) {
            oRm.addStyle('width', oControl.getWidth());
        }
        oRm.writeStyles();
        oRm.write('>');
        var oContent = oControl._renderingContent ? oControl._renderingContent() : oControl._getCompositeAggregation();
        if (oContent) {
            oRm.renderControl(oContent);
        }
        oRm.write('</div>');
        Measurement.end(oControl.getId() + '---renderControl');
        Log.debug('Stop rendering \'' + oControl.sId, sXMLComposite);
    }
}, XMLCompositeMetadata);
XMLComposite.prototype.byId = function (sId) {
    return sap.ui.getWCCore().byId(Fragment.createId(this.getId(), sId));
};
XMLComposite.prototype._getManagedObjectModel = function () {
    if (!this._oManagedObjectModel) {
        this._oManagedObjectModel = new ManagedObjectModel(this);
    }
    return this._oManagedObjectModel;
};
XMLComposite.prototype.getSuppressInvalidateAggregation = function (sName, bSuppressInvalidate) {
    var oMetadata = this.getMetadata(), oAggregation = oMetadata.getAggregation(sName) || oMetadata.getAllPrivateAggregations()[sName];
    if (!oAggregation) {
        return true;
    }
    bSuppressInvalidate = oMetadata._suppressInvalidate(oAggregation, bSuppressInvalidate);
    return bSuppressInvalidate;
};
XMLComposite.prototype.setProperty = function (sName, oValue, bSuppressInvalidate) {
    var oMetadata = this.getMetadata(), oProperty = oMetadata.getProperty(sName);
    if (!oProperty) {
        return this;
    }
    bSuppressInvalidate = this.getMetadata()._suppressInvalidate(oProperty, bSuppressInvalidate);
    return Control.prototype.setProperty.apply(this, [
        sName,
        oValue,
        bSuppressInvalidate
    ]);
};
XMLComposite.prototype.setAggregation = function (sName, oObject, bSuppressInvalidate) {
    return Control.prototype.setAggregation.apply(this, [
        sName,
        oObject,
        this.getSuppressInvalidateAggregation(sName, bSuppressInvalidate)
    ]);
};
XMLComposite.prototype.addAggregation = function (sName, oObject, bSuppressInvalidate) {
    return Control.prototype.addAggregation.apply(this, [
        sName,
        oObject,
        this.getSuppressInvalidateAggregation(sName, bSuppressInvalidate)
    ]);
};
XMLComposite.prototype.insertAggregation = function (sName, oObject, iIndex, bSuppressInvalidate) {
    return Control.prototype.insertAggregation.apply(this, [
        sName,
        oObject,
        iIndex,
        this.getSuppressInvalidateAggregation(sName, bSuppressInvalidate)
    ]);
};
XMLComposite.prototype.removeAggregation = function (sName, oObject, bSuppressInvalidate) {
    return Control.prototype.removeAggregation.apply(this, [
        sName,
        oObject,
        this.getSuppressInvalidateAggregation(sName, bSuppressInvalidate)
    ]);
};
XMLComposite.prototype.removeAllAggregation = function (sName, bSuppressInvalidate) {
    return Control.prototype.removeAllAggregation.apply(this, [
        sName,
        this.getSuppressInvalidateAggregation(sName, bSuppressInvalidate)
    ]);
};
XMLComposite.prototype.destroyAggregation = function (sName, bSuppressInvalidate) {
    return Control.prototype.destroyAggregation.apply(this, [
        sName,
        this.getSuppressInvalidateAggregation(sName, bSuppressInvalidate)
    ]);
};
XMLComposite.prototype.updateAggregation = function (sName, bSuppressInvalidate) {
    var oAggregation = this.getMetadata().getAggregation(sName);
    if (oAggregation && oAggregation.type === 'TemplateMetadataContext') {
        this.invalidate();
        return;
    }
    Control.prototype.updateAggregation.apply(this, arguments);
};
XMLComposite.prototype.setVisible = function (bVisible) {
    this.setProperty('visible', bVisible);
    if (this.getParent()) {
        this.getParent().invalidate();
    }
    return this;
};
XMLComposite.prototype._destroyCompositeAggregation = function () {
    var oContent = this._getCompositeAggregation();
    if (oContent) {
        oContent.destroy('KeepDom');
    }
    return this;
};
XMLComposite.prototype.updateBindings = function () {
    if (this._bIsCreating) {
        return;
    }
    var oResult = Control.prototype.updateBindings.apply(this, arguments);
    for (var n in this.mBindingInfos) {
        var oAggregation = this.getMetadata().getAggregation(n);
        if (oAggregation && oAggregation.multiple && !oAggregation._doesNotRequireFactory && this.isBound(n) && !this.getBinding(n)) {
            this[oAggregation._sDestructor]();
        }
    }
    return oResult;
};
XMLComposite.prototype._getCompositeAggregation = function () {
    var sCompositeName = this.getMetadata().getCompositeAggregationName();
    return this.getAggregation(sCompositeName);
};
XMLComposite.prototype._setCompositeAggregation = function (oNewContent) {
    var sCompositeName = this.getMetadata().getCompositeAggregationName();
    this._destroyCompositeAggregation();
    if (!this._oManagedObjectModel) {
        this._getManagedObjectModel();
    }
    if (Array.isArray(oNewContent)) {
        this.setAggregation(sCompositeName, null);
        return;
    }
    if (oNewContent) {
        if (!oNewContent.enhanceAccessibilityState) {
            oNewContent.enhanceAccessibilityState = function (oElement, mAriaProps) {
                this.enhanceAccessibilityState(oElement, mAriaProps);
            }.bind(this);
        }
        oNewContent.bindObject('$' + this.alias + '>/');
        oNewContent.setModel(this._oManagedObjectModel, '$' + this.alias);
        var oResourceModel = this._getResourceModel();
        if (oResourceModel) {
            oNewContent.setModel(oResourceModel, '$' + this.alias + '.i18n');
        }
    }
    this.setAggregation(sCompositeName, oNewContent);
};
XMLComposite.mResourceModels = {};
XMLComposite.getLibraryResourceModel = function (sLibraryName) {
    var oLibraryResourceModel = XMLComposite.mResourceModels[sLibraryName];
    if (!oLibraryResourceModel) {
        oLibraryResourceModel = new ResourceModel({
            bundleName: sLibraryName + '.messagebundle',
            async: true
        });
        XMLComposite.mResourceModels[sLibraryName] = oLibraryResourceModel;
    }
    return oLibraryResourceModel;
};
XMLComposite.prototype._getResourceModel = function () {
    if (this.resourceModel) {
        return this.resourceModel;
    }
    if (this.messageBundle) {
        this.resourceModel = new ResourceModel({
            bundleName: this.messageBundle,
            async: true
        });
        return this.resourceModel;
    } else {
        this.sLibraryName = this.sLibraryName || this.getMetadata().getLibraryName();
        if (this.sLibraryName) {
            return XMLComposite.getLibraryResourceModel(this.sLibraryName);
        }
    }
};
XMLComposite.prototype.getResourceBundle = function () {
    var oResourceModel = this._getResourceModel();
    return oResourceModel ? oResourceModel.getResourceBundle() : null;
};
XMLComposite.prototype.destroy = function () {
    Control.prototype.destroy.apply(this, arguments);
    if (this.resourceModel) {
        this.resourceModel.destroy();
    }
    if (this._oManagedObjectModel) {
        this._oManagedObjectModel.destroy();
    }
};
XMLComposite.prototype._initCompositeSupport = function (mSettings) {
    var oMetadata = this.getMetadata(), oFragmentContent = oMetadata._fragment, sAggregationName = oMetadata.getCompositeAggregationName();
    this._destroyCompositeAggregation();
    if (mSettings && sAggregationName && mSettings[sAggregationName]) {
        var oNode = mSettings[sAggregationName];
        if (oNode.localName === 'FragmentDefinition') {
            oFragmentContent = oNode;
            delete mSettings[sAggregationName];
        }
    }
    this._setCompositeAggregation(sap.ui.xmlfragment({
        sId: this.getId(),
        fragmentContent: oFragmentContent,
        oController: this
    }));
    this._bIsInitialized = true;
};
XMLComposite.prototype.enhanceAccessibilityState = function (oElement, mAriaProps) {
    var oParent = this.getParent();
    if (oParent && oParent.enhanceAccessibilityState) {
        return oParent.enhanceAccessibilityState(this, mAriaProps);
    }
    return mAriaProps;
};
XMLComposite.prototype.getFocusDomRef = function () {
    var oContent = this._renderingContent ? this._renderingContent() : this._getCompositeAggregation();
    return oContent.getFocusDomRef();
};
XMLComposite.prototype.getFocusInfo = function () {
    var oContent = this._renderingContent ? this._renderingContent() : this._getCompositeAggregation();
    return oContent.getFocusInfo();
};
XMLComposite.prototype.getIdForLabel = function () {
    var oContent = this._renderingContent ? this._renderingContent() : this._getCompositeAggregation();
    return oContent.getIdForLabel();
};
export default XMLComposite;
