// @ui5/webcomponents-core/dist/sap/ui/core/search/OpenSearchProvider.js
import library from '../library.js';
import SearchProvider from './SearchProvider.js';
import Log from '../../../base/Log.js';
import encodeURL from '../../../base/security/encodeURL.js';
import jQuery from '../../thirdparty/jquery.js';
var OpenSearchProvider = SearchProvider.extend('sap.ui.core.search.OpenSearchProvider', {
    metadata: {
        library: 'sap.ui.core',
        properties: {
            suggestUrl: {
                type: 'sap.ui.core.URI',
                group: 'Misc',
                defaultValue: null
            },
            suggestType: {
                type: 'string',
                group: 'Misc',
                defaultValue: 'json'
            }
        }
    }
});
OpenSearchProvider.prototype.suggest = function (sValue, fCallback) {
    var sUrl = this.getSuggestUrl();
    if (!sUrl) {
        return;
    }
    sUrl = sUrl.replace('{searchTerms}', encodeURL(sValue));
    var sType = this.getSuggestType();
    var fSuccess;
    if (sType && sType.toLowerCase() === 'xml') {
        sType = 'xml';
        fSuccess = function (data) {
            var jXMLDocument = jQuery(data);
            var jItems = jXMLDocument.find('Text');
            var aSuggestions = [];
            jItems.each(function () {
                aSuggestions.push(jQuery(this).text());
            });
            fCallback(sValue, aSuggestions);
        };
    } else {
        sType = 'json';
        fSuccess = function (data) {
            fCallback(sValue, data[1]);
        };
    }
    jQuery.ajax({
        url: sUrl,
        dataType: sType,
        success: fSuccess,
        error: function (XMLHttpRequest, textStatus, errorThrown) {
            Log.fatal('The following problem occurred: ' + textStatus, XMLHttpRequest.responseText + ',' + XMLHttpRequest.status);
        }
    });
};
export default OpenSearchProvider;
