// @ui5/webcomponents-core/dist/sap/ui/core/routing/async/Target.js
import Log from '../../../../base/Log.js';
import ComponentContainer from '../../ComponentContainer.js';
import coreLib from '../../library.js';
export default {
    display: function (vData) {
        var oSequencePromise = Promise.resolve();
        return this._display(vData, oSequencePromise).catch(function (oViewInfo) {
            if (oViewInfo instanceof Error) {
                return Promise.reject(oViewInfo);
            } else {
                return oViewInfo;
            }
        });
    },
    _display: function (vData, oSequencePromise, oTargetCreateInfo) {
        if (this._oParent) {
            oSequencePromise = this._oParent._display(vData, oSequencePromise, oTargetCreateInfo);
        }
        return this._place(vData, oSequencePromise, oTargetCreateInfo);
    },
    _place: function (vData, oSequencePromise, oTargetCreateInfo) {
        if (vData instanceof Promise) {
            oTargetCreateInfo = oSequencePromise;
            oSequencePromise = vData;
            vData = undefined;
        }
        var oOptions = this._oOptions, that = this, oObject, sName, oCreateOptions, sErrorMessage, pLoaded;
        if (oOptions.name && oOptions.type) {
            sName = this._getEffectiveObjectName(oOptions.name);
            switch (oOptions.type) {
            case 'View':
                oCreateOptions = {
                    name: sName,
                    type: oOptions.viewType,
                    id: oOptions.id,
                    async: true
                };
                break;
            case 'Component':
                oCreateOptions = Object.assign({}, oOptions.options || {}, {
                    name: sName,
                    id: oOptions.id
                });
                break;
            default:
                throw new Error('The given type ' + oOptions.type + ' isn\'t support by sap.ui.core.routing.Target');
            }
            oObject = this._oCache._get(oCreateOptions, oOptions.type, this._bUseRawViewId, oTargetCreateInfo);
            if (!(oObject instanceof Promise)) {
                if (oObject.isA('sap.ui.core.mvc.View')) {
                    pLoaded = oObject.loaded();
                } else {
                    pLoaded = Promise.resolve(oObject);
                }
            } else {
                pLoaded = oObject;
            }
            oSequencePromise = oSequencePromise.then(function (oParentInfo) {
                return pLoaded.then(function (oObject) {
                    if (oObject.isA('sap.ui.core.UIComponent')) {
                        var oRouter = oObject.getRouter();
                        if (oRouter && oRouter.isStopped()) {
                            oRouter.initialize();
                        }
                    }
                    return {
                        object: oObject,
                        parentInfo: oParentInfo || {}
                    };
                }, function (sErrorMessage) {
                    return Promise.reject({
                        name: oOptions._name,
                        error: sErrorMessage
                    });
                });
            }).then(function (oViewInfo) {
                var vValid = that._isValid(oViewInfo.parentInfo);
                oObject = oViewInfo.object;
                if (oObject.isA('sap.ui.core.mvc.View')) {
                    that._bindTitleInTitleProvider(oObject);
                    that._addTitleProviderAsDependent(oObject);
                }
                if (vValid !== true) {
                    sErrorMessage = vValid;
                    return that._refuseInvalidTarget(oOptions._name, sErrorMessage);
                }
                var oViewContainingTheControl = oViewInfo.parentInfo.view, oControl = oViewInfo.parentInfo.control, pContainerControl = Promise.resolve(oControl);
                if (oViewContainingTheControl && oViewContainingTheControl.isA('sap.ui.core.UIComponent')) {
                    oViewContainingTheControl = oViewContainingTheControl.getRootControl();
                }
                if (!oViewContainingTheControl && oOptions.rootView) {
                    oViewContainingTheControl = sap.ui.getWCCore().byId(oOptions.rootView);
                    if (!oViewContainingTheControl) {
                        sErrorMessage = 'Did not find the root view with the id ' + oOptions.rootView;
                        return that._refuseInvalidTarget(oOptions._name, sErrorMessage);
                    }
                }
                if (oOptions.controlId) {
                    if (oViewContainingTheControl) {
                        pContainerControl = oViewContainingTheControl.loaded().then(function (oContainerView) {
                            return oContainerView.byId(oOptions.controlId);
                        });
                    }
                    pContainerControl = pContainerControl.then(function (oContainerControl) {
                        if (!oContainerControl) {
                            oContainerControl = sap.ui.getWCCore().byId(oOptions.controlId);
                        }
                        if (!oContainerControl) {
                            sErrorMessage = 'Control with ID ' + oOptions.controlId + ' could not be found';
                            return that._refuseInvalidTarget(oOptions._name, sErrorMessage);
                        } else {
                            return oContainerControl;
                        }
                    }).catch(function () {
                        sErrorMessage = 'Something went wrong during loading the root view with id ' + oOptions.rootView;
                        return that._refuseInvalidTarget(oOptions._name, sErrorMessage);
                    });
                }
                return pContainerControl;
            }).then(function (oContainerControl) {
                var oComponent, sComponentContainerId, fnOriginalExit;
                if (oContainerControl.error) {
                    return oContainerControl;
                }
                if (oObject.isA('sap.ui.core.UIComponent')) {
                    oComponent = oObject;
                    sComponentContainerId = oComponent.getId() + '-container';
                    oObject = sap.ui.getWCCore().byId(sComponentContainerId);
                    if (!oObject) {
                        var oContainerOptions = Object.assign({
                            component: oComponent,
                            height: '100%',
                            width: '100%',
                            lifecycle: sap.ui.core.ComponentLifecycle.Application
                        }, oOptions.containerOptions);
                        oObject = new ComponentContainer(sComponentContainerId, oContainerOptions);
                        fnOriginalExit = oComponent.exit;
                        oComponent.exit = function () {
                            if (fnOriginalExit) {
                                fnOriginalExit.apply(this);
                            }
                            oObject.destroy();
                        };
                    }
                }
                that._beforePlacingViewIntoContainer({
                    container: oContainerControl,
                    view: oObject,
                    data: vData
                });
                var oAggregationInfo = oContainerControl.getMetadata().getJSONKeys()[oOptions.controlAggregation];
                if (!oAggregationInfo) {
                    sErrorMessage = 'Control ' + oOptions.controlId + ' does not have an aggregation called ' + oOptions.controlAggregation;
                    return that._refuseInvalidTarget(oOptions._name, sErrorMessage);
                }
                if (oOptions.clearControlAggregation === true) {
                    oContainerControl[oAggregationInfo._sRemoveAllMutator]();
                }
                Log.info('Did place the ' + oOptions.type.toLowerCase() + ' \'' + sName + '\' with the id \'' + oObject.getId() + '\' into the aggregation \'' + oOptions.controlAggregation + '\' of a control with the id \'' + oContainerControl.getId() + '\'', that);
                oContainerControl[oAggregationInfo._sMutator](oObject);
                that.fireDisplay({
                    view: oObject.isA('sap.ui.core.mvc.View') ? oObject : undefined,
                    object: oObject,
                    control: oContainerControl,
                    config: that._oOptions,
                    data: vData
                });
                return {
                    name: oOptions._name,
                    view: oObject,
                    control: oContainerControl
                };
            });
        } else {
            oSequencePromise = oSequencePromise.then(function () {
                return { name: oOptions._name };
            });
        }
        return oSequencePromise;
    },
    _isValid: function (oParentInfo) {
        var oOptions = this._oOptions, oControl = oParentInfo && oParentInfo.control, bHasTargetControl = oControl || oOptions.controlId, bIsValid = true, sLogMessage = '';
        if (!bHasTargetControl) {
            sLogMessage = 'The target ' + oOptions._name + ' has no controlId set and no parent so the target cannot be displayed.';
            bIsValid = false;
        }
        if (!oOptions.controlAggregation) {
            sLogMessage = 'The target ' + oOptions._name + ' has a control id or a parent but no \'controlAggregation\' was set, so the target could not be displayed.';
            bIsValid = false;
        }
        if (sLogMessage) {
            Log.error(sLogMessage, this);
        }
        return bIsValid || sLogMessage;
    },
    _refuseInvalidTarget: function (sName, sMessage) {
        if (sMessage) {
            Log.error(sMessage, this);
        }
        return {
            name: sName,
            error: sMessage
        };
    }
};
