// @ui5/webcomponents-core/dist/sap/ui/debug/Highlighter.js
import jQuery from '../thirdparty/jquery.js';
import uid from '../../base/util/uid.js';
import '../dom/jquery/rect.js';
var Highlighter = function (sId, bFilled, sColor, iBorderWidth) {
    this.sId = sId || uid();
    this.bFilled = bFilled == true;
    this.sColor = sColor || 'blue';
    if (isNaN(iBorderWidth)) {
        this.iBorderWidth = 2;
    } else if (iBorderWidth <= 0) {
        this.iBorderWidth = 0;
    } else {
        this.iBorderWidth = iBorderWidth;
    }
};
Highlighter.prototype.highlight = function (oDomRef) {
    if (!oDomRef || !oDomRef.parentNode) {
        return;
    }
    var oHighlightRect = this.sId ? window.document.getElementById(this.sId) : null;
    if (!oHighlightRect) {
        oHighlightRect = oDomRef.ownerDocument.createElement('DIV');
        oHighlightRect.setAttribute('id', this.sId);
        oHighlightRect.style.position = 'absolute';
        oHighlightRect.style.border = this.iBorderWidth + 'px solid ' + this.sColor;
        oHighlightRect.style.display = 'none';
        oHighlightRect.style.margin = '0px';
        oHighlightRect.style.padding = '0px';
        if (this.bFilled) {
            oHighlightRect.innerHTML = '<div style=\'background-color:' + this.sColor + ';opacity:0.2;filter:progid:DXImageTransform.Microsoft.Alpha(opacity=20);height:100%;width:100%\'>&nbsp;</div>';
        }
        oDomRef.ownerDocument.body.appendChild(oHighlightRect);
    }
    var oRect = jQuery(oDomRef).rect();
    oHighlightRect.style.top = oRect.top - this.iBorderWidth + 'px';
    oHighlightRect.style.left = oRect.left - this.iBorderWidth + 'px';
    oHighlightRect.style.width = oRect.width + 'px';
    oHighlightRect.style.height = oRect.height + 'px';
    oHighlightRect.style.display = 'block';
};
Highlighter.prototype.hide = function () {
    var oHighlightRect = this.sId ? window.document.getElementById(this.sId) : null;
    if (!oHighlightRect) {
        return;
    }
    oHighlightRect.style.display = 'none';
};
export default Highlighter;
